cc/main.cc

```diff
@@ -464,46 +467,244 @@ class SelfPlayer {
+  // of a work queue where some threads exit early.
+  class Barrier {
+   public:
+    explicit Barrier(size_t count)
+        : count_(count), num_waiting_(0), generation_(0) {}
+
+    void Wait() {
+      absl::MutexLock lock(&mutex_);
+      if (++num_waiting_ == count_) {
+        IncrementGeneration();
+      } else {
+        auto generation = generation_;
+        while (generation != generation_) {
+          cond_var_.Wait(&mutex_);
+        }
+      }
+    }
+
+    void DecrementCount() {
+      absl::MutexLock lock(&mutex_);
+      if (num_waiting_ == --count_) {
+        IncrementGeneration();
+      }
+    }
+
+   private:
+    void IncrementGeneration() EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
+      ++generation_;
+      num_waiting_ = 0;
+      cond_var_.SignalAll();
+    }
+
+    absl::Mutex mutex_;
+    absl::CondVar cond_var_;
+    size_t count_ GUARDED_BY(&mutex_);
+    size_t num_waiting_ GUARDED_BY(&mutex_);
+    size_t generation_ GUARDED_BY(&mutex_);
+  };
+
+  // References a pointer to an actual DualNet. Allows updating the pointer
+  // after the MctsPlayer has been constructed.
+  class WrappedDualNet : public DualNet {
+   public:
+    explicit WrappedDualNet(const std::unique_ptr<DualNet>& dual_net)
+        : dual_net_(dual_net) {}
+
+   private:
+    void RunMany(std::vector<const BoardFeatures*> features,
+                 std::vector<Output*> outputs, std::string* model) override {
+      dual_net_->RunMany(std::move(features), std::move(outputs), model);
+    };
+
+    const std::unique_ptr<DualNet>& dual_net_;
+  };
+
+  struct Model {
+    std::string name;
+    std::unique_ptr<DualNetFactory> factory;
+    std::atomic<int> black_wins;
+    std::atomic<int> white_wins;
+  };
+
+ public:
+  void Run() {
+    auto start_time = absl::Now();
+
+    auto create_model = [](const std::string& model_path) {
+      auto model = absl::make_unique<Model>();
+      model->name = static_cast<std::string>(file::Stem(model_path));
+      model->factory = NewDualNetFactory(model_path);
+      model->black_wins = 0;
+      model->white_wins = 0;
+      return model;
+    };
+
+    auto prev_model = create_model(FLAGS_model);
+    auto cur_model = create_model(FLAGS_model_two);
+
+    std::cerr << "DualNet factories created from " << FLAGS_model << "\n  and "
+              << FLAGS_model_two << " in "
+              << absl::ToDoubleSeconds(absl::Now() - start_time) << " sec."
+              << std::endl;
+
+    ParseMctsPlayerOptionsFromFlags(&options_);
+    options_.inject_noise = false;
+    options_.soft_pick = false;
+    options_.random_symmetry = true;
+
+    int num_games = FLAGS_parallel_games;
+    barrier_ = absl::make_unique<Barrier>(num_games);
+
+    for (int i = 0; i < num_games;) {
+      threads_.emplace_back(std::bind(&Evaluator::ThreadRun, this, i,
+                                      cur_model.get(), prev_model.get()));
+      ++i;
+      if (FLAGS_even && i < num_games) {
```

Comment:
Instead of putting the "even" logic here in a separate lambda, why make call barrier_->Wait() at the top of ThreadRun for only the threads with odd ids? The two Wait calls would be closer in the code that way too, which aids readability.